packages/instantsearch.js/src/lib/server.ts

```diff
@@ -37,17 +57,27 @@ export function waitForResults(search: InstantSearch): Promise<void> {
 /**
  * Walks the InstantSearch root index to construct the initial results.
  */
-export function getInitialResults(rootIndex: IndexWidget): InitialResults {
+export function getInitialResults(
+  rootIndex: IndexWidget,
+  /**
+   * Search parameters sent to the search client, usually
+   * returned by `waitForResults()`.
+   */
+  requestParamsList?: RequestParams[]
+): InitialResults {
   const initialResults: InitialResults = {};
 
   walkIndex(rootIndex, (widget) => {
     const searchResults = widget.getResults();
     if (searchResults) {
-      initialResults[widget.getIndexId()] = {
+      const indexId = widget.getIndexId();
+      const requestParams = requestParamsList?.shift();
```

Comment:
should we keep track of the index and use the item at index `i` instead of mutating? Not sure if it's a good idea that `getInitialResults` effectively can only be called once (also let's add a test for calling it multiple times)